### PR TITLE
Streaming Server

### LINKED ISSUE #1653: In ASGI, headers is output as body
simple test code
```Python3
@app.get('')
async def test(request):
    return await file_stream(...)
```

The HTTP response will output twice headers.
One of them is in the headers (✅), the other time in the body (❎)

### LINKED ISSUE #1722: curl is not agree wit return in chunk mode
for uploading large file  > 200 MB I write:

@app.post('/data', stream=True)
async def handler(_request):
    async def streaming(_response):
        total_size = 0
        while True:
            body = await _request.stream.read()
            if body is None:
                break
            total_size += len(body)
            print("body " + str(len(body)) + "/" + str(total_size))
    return response.stream(streaming)
```
And send the file with 
```
curl  -F 'file=@Totally_Spies.mp4' -H 'transfer-encoding:chunked' 127.0.0.1:15080/data -X POST -O
```

The curl send #95MB befor stoping with the error:
```
curl: (56) Illegal or missing hexadecimal sequence in chunked-encoding
```

I do not understand what is wrong!

### LINKED ISSUE #1730: ASGI Streaming response contains extra content
I have some code that makes an http request (aiohttp) to an separate service to retrieve a file and then stream that file back.  It is essentially acting as a proxy for this request.  However, the streaming portion ends up wrapping some additional bytes of data around the body that I don't quite understand.

Here is a test case:

```
from sanic import Sanic
from sanic.response import file_stream
from pathlib import Path
import aiofiles
from aiofiles import os as async_os
from sanic.response import stream
import aiohttp

app = Sanic()

@app.route("/", methods=["GET"])
async def test(request):
  tf=__file__
  file_stat = await async_os.stat(tf)
  headers = {"Content-Length": file_stat.st_size,"Content-Type": "application/octet-stream"}
  return await file_stream(tf,headers=headers,chunked=False)

@app.route("/test2",methods=["GET"])
async def test2(request):
  session = aiohttp.ClientSession()
  resp = await session.get('http://localhost:%s/'%request.server_port)

  async def stream_file(response):
    while True:
      chunk = await resp.content.read()
      if not chunk:
        break
      await response.write(chunk)

  return stream(stream_file)

```
I run this code with 'uvicorn test:app --port 9000'

There are two endpoints / and /test2:
- The / request simply returns a file (the actual code in this example). curl "http://localhost:9000/" verifies the results are correct
- The /test2 call makes an aiohttp call to the / endpoint and then tries to stream the response.  I expect the response to be the same as the / endpoint.

The actual response looks like:
```
$ curl http://localhost:9000/test2
33d
from sanic import Sanic
from sanic.response import file_stream
from pathlib import Path
...TRUNCATED MIDDLE OF OUTPUT
  return stream(stream_file)
0
```
That '33d' and '0' should not be there, but I don't see where they are coming from.  Am I passing the stream along improperly?

The environment is OSX 10.15.1

One note is that  I have applied the small snippet of code I described in #1653 in order to make the initial file stream work.

### LINKED ISSUE #1749: Sanic 19.12 breaks some middleware usages
The commit _GIT-37: fix blueprint middleware application_ (https://github.com/huge-success/sanic/commit/a6077a179067812766293f9ab4391b5e1499a9c6#diff-c4444a35dbd5e37ee480b0e8888e0880R939) changes how the middleware are evaluated. 

Before this commit the middleware where evaluated before the route was determined. This allowed for example to rewrite to requested url. In my case I was using that to remove a url-prefix that is used for routing purposes.

I do think this is a legitimate use case for middlewares. According to the documentation:

> Middleware are functions which are executed before or after requests to the server. They can be used to modify the request to or response from user-defined handler functions.

modifying the requested url should be possible.

### LINKED ISSUE #1785: Error handlers are synchronous
**Describe the bug**
The [documentation](https://sanic.readthedocs.io/en/latest/sanic/exceptions.html) says that error handlers can be asynchronous.
However, asynchronous error handlers do not seem to be supported everywhere.

**Code snippet**
```py
@app.exception(Exception)
async def handle_base_exception(_request: Request, exception: Exception):
    return response.json("too bad")
```

**Expected behavior**
Then, for example when I exceed the `REQUEST_MAX_SIZE` limit, I would expect to receive `"too bad"` as a response.

Unfortunately, sanic contains code like https://github.com/huge-success/sanic/blob/91f6abaa81248189fbcbdf685e8bdcbb7846609f/sanic/server.py#L577-L579

which expects the error handler to be synchronous.

**Actual behavior**

Here are the (verbose) logs of the actual results:

```
2020-02-17 19:03:15,514:INFO:
[2020-02-17 19:03:24 +0100] [37783] [ERROR] Transport closed @ ('127.0.0.1', 58698) and exception experienced during error handling
[2020-02-17 19:03:24 +0100] - (sanic.access)[INFO][UNKNOWN]: nil  0 -1
2020-02-17 19:03:24,145:ERROR:Transport closed @ ('127.0.0.1', 58698) and exception experienced during error handling
2020-02-17 19:03:24,145:INFO:
.../venv/lib/python3.7/site-packages/sanic/server.py:297: RuntimeWarning: coroutine 'handle_base_exception' was never awaited
  self.write_error(PayloadTooLarge("Payload Too Large"))
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

### LINKED ISSUE #1804: Why define an exception handler needed async?
with tests/test_exceptions_handler.py define an exception handler like this:

> @exception_handler_app.exception(NotFound, ServerError)
> def handler_exception(request, exception):
>     return text("OK")

but in docs [https://sanic.readthedocs.io/en/latest/sanic/exceptions.html](https://sanic.readthedocs.io/en/latest/sanic/exceptions.html)

> @app.exception(NotFound)
> async def ignore_404s(request, exception):
>     return text("Yep, I totally found the page: {}".format(request.url))

what is the difference? And is async def needed?

### LINKED ISSUE #1824: How to increase timeout on specific method?
Is there any option to increase timeout and request size on the top of method such as decorators?
Sometimes we need to increase our time of request such as uploading files.
Also, I asked a question on [Stackoverflow](https://stackoverflow.com/questions/60915546/how-to-increase-timeout-while-uploading-on-the-sanic-framework) about this issue.

### LINKED ISSUE #1875: Response timeout error from all routes after returning several timeouts from a specific route
**Response timeout error**
Return "Response timeout" error from all routes after returning several timeouts from a specific route.

**Codes**
Set response timeout to 5 sec in config:
`RESPONSE_TIMEOUT = 5`

Define funcs:
```
@app.get('test')
async def test_func(request):
    """Timeout func"""
    await asyncio.sleep(10)
    ...
@app.get('test2')
async def test_func2(request):
    """Func without timeout"""
    return response.text('OK')
```
Make several timeout responses:
```
for i in range(20):
    resp = requests.get('/test')
```
Then if I request all other endpoints, it returns "Response timeout" error:
`resp = requests.get('/test2')`

**Environment:**
 - OS: Ubuntu
 - Python: 3.8.2
 - Sanic: 20.3.0

### LINKED ISSUE #1988: Handling of safe methods with body
**Is your feature request related to a problem? Please describe.**

Currently, Sanic will allow a request body to be accepted (and parsed) even on safe methods (GET, HEAD, DELETE, OPTIONS). This is _technically_ accurate [since the specs do not require](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET) that a server disallow this behavior.

According to Mozilla:

> ... while not prohibited by the specification, the semantics are undefined.

It is is, however, somewhat counter-intuitive.

**Describe the solution you'd like**

Add a flag (`ignore_body`) to the save method decorators that defaults to `True`. If it receives a safe request with a body while `ignore_body == True`, then it should **not** parse the body.

### PR DESCRIPTION
**Pull request by Tronic**
Moved from https://github.com/huge-success/sanic/pull/1791

A plain Python reimplementation of Sanic's built-in server and related restructuring also in ASGI backend and in response handling. Includes first-class support for streaming, improved performance and better ground for further development.

By the sheer volume of it, it also brings some breaking changes like the removal of custom protocol support but typical applications that do not do anything special with the server's internals should remain mostly unaffected.

Installation:
```
pip install git+https://github.com/huge-success/sanic@streaming
```

Issues of current git master, resolved in this:
Fix #1653
Fix #1722
Fix #1730
Fix #1749 
Fix #1785
Fix #1790
Fix #1804
Fix #1824
Fix #1875
Fix #1988

### CODE REVIEW

**Comment by ahopkins** [comment]
I am sure there are already a number of conflicts here. Before it gets too wildly out of sync, do you want to try to do a rebase (or merge, not picky) to bring this back current to `master`? I am thinking that I want to keep this at the forefront so it is ready to merge asap after December's LTS, with some of the other major changes that are waiting to follow after this.

**Comment by Tronic** [comment]
@ahopkins Help with that would be nice. I'm busy with work and academia pretty much 7 days every week until Nov/Dec, also you are probably better familiar with all the patches merged to master meanwhile. I believe it would be easier to merge them to this branch than rebase the whole thing.

**Comment by ahopkins** [comment]
@Tronic. Sounds good. :metal: 

After the release this week ...

**Comment by ahopkins** [comment]
I merged in recent changes to master. I did not spend any time on the tests. Looks like about 12 need some attention. I can take a look at these this week.

**Comment by ahopkins** [comment]
@huge-success/sanic-core-devs Please take a look at this one. I'd like to get this merged by the end of the weekend. Even if it is not 100% perfect and release worthy, I want to get this merged because of the large number of changes that it entails. Having it merged into `master` will allow us to continue making other PRs without resolving too many conflicts. Release it not set until March, but I want to make sure there is **plenty** of time to test this in `master`.

**Comment by ahopkins** [comment]
This pull request has been mentioned on **Sanic Community Discussion**. There might be relevant details there:

https://community.sanicframework.org/t/new-streaming-api-and-other-major-changes/518/15

**Comment by ahopkins** [comment]
This pull request has been mentioned on **Sanic Community Discussion**. There might be relevant details there:

https://community.sanicframework.org/t/2021-release-managers-and-updates/781/1

File: sanic/asgi.py [hunk 1]
```diff
@@ -257,161 +241,57 @@ async def create(
                 instance.transport,
                 sanic_app,
             )
+            instance.request.stream = instance
+            instance.request_body = True  # FIXME: Use more_body?
```
Comment (ahopkins) [inline sanic/asgi.py:245]: Needs resolution.

File: sanic/request.py [hunk 2]
```diff
@@ -135,36 +110,43 @@ def __repr__(self):
...
-        """.. deprecated:: 20.3
-        To be removed in 21.3"""
-        self.body.append(data)
-
-    def body_finish(self):
-        """.. deprecated:: 20.3
-        To be removed in 21.3"""
-        self.body = b"".join(self.body)
+    async def respond(
+        self, response=None, *, status=200, headers=None, content_type=None
```
Comment (ahopkins) [inline sanic/request.py:114]: Need to take a closer look to confirm that this is actually the correct place for this.

File: sanic/response.py [hunk 3]
```diff
@@ -28,50 +27,51 @@ def _encode_body(self, data):
...
-        else:
-            self.headers["Connection"] = "close"
+    @property
+    def processed_headers(self):
+        """Obtain a list of header tuples encoded in bytes for sending.
 
-        if self.status in (304, 412):
+        Add and remove headers based on status and content_type.
+        """
+        # TODO: Make a blacklist set of header names and then filter with that
```
Comment (ahopkins) [inline sanic/response.py:42]: Resolve.

File: sanic/response.py [hunk 4]
```diff
@@ -81,63 +81,28 @@ def __init__(
         status=200,
         headers=None,
         content_type="text/plain; charset=utf-8",
-        chunked=True,
+        chunked="deprecated",
```
Comment (ahopkins) [inline sanic/response.py:84]: Remove?

File: sanic/response.py [hunk 5]
```diff
@@ -224,7 +173,15 @@ def text(
         raise TypeError(
             f"Bad body type. Expected str, got {type(body).__name__})"
         )
-
+    # Type conversions are deprecated and quite b0rked but still supported for
+    # text() until applications get fixed. This try-except should be removed.
+    try:
+        # Avoid repr(body).encode() b0rkage for body that is already encoded.
+        # memoryview used only to test bytes-ishness.
+        with memoryview(body):
```
Comment (ahopkins) [inline sanic/response.py:181]: Let's benchmark this.

**Review by ahopkins — CHANGES_REQUESTED** [review]
I'll handle these. When done, I want to merge. @sanic-org/sanic-core-devs any other thoughts before I do so?

**Review by ahopkins — APPROVED** [review]

**Comment by ashleysommer** [comment]
Just looking over these merged changes now, to learn how Sanic works at a low level now.

I have noticed it looks like there is a possibility for the Request Middleware to run twice on a request.
Both the `app.handle_request` and `app.handle_exception` methods execute the Request middleware.

If the main handler encounters an exception _after_ running the middleware, then the exception handler will execute and run the request middleware _again_.

**Comment by ahopkins** [comment]
Yeah. That is the purpose of it.  I think eventually I want to get rid of it and not have to carry that state trigger (there is not a similar one for response).

**Comment by ahopkins** [comment]
This pull request has been mentioned on **Sanic Community Discussion**. There might be relevant details there:

https://community.sanicframework.org/t/highly-different-cpu-usage-when-proxying-m2tp-udp-unicast-stream-to-a-streamed-sanic-response/795/10

**Pull request merged**